fio/lib/src/utils.ts

```diff
@@ -0,0 +1,281 @@
+    account_name: "fio.address",
+    code_hash:
+      "72c258fbcb0b328960bdbc4fc521d76cc53e5517bb5f5f2bf991d08030ddae6b",
+  };
+
+  const abiMap = new Map<string, FioTokenAbi>();
+  abiMap.set(account, fioTokenAbi);
+
+  const abiProvider: AbiProvider = {
+    getRawAbi: async () => {
+      const rawAbi = abiMap.get(account);
+      if (!rawAbi) throw new Error("Missing abi for " + account);
+      const abi = base64ToBinary(rawAbi.abi);
+      const binaryAbi = {
+        accountName: rawAbi.account_name,
+        abi: abi,
+      };
+      return binaryAbi;
+    },
+  };
+
+  return { fioTokenAbi, abiProvider };
+};
+
+/**
+ * Gets chain data.
+ * @returns The chain data.
+ */
+const getChainData = async (fioBaseUrl: string): Promise<ChainData> => {
+  const chainInfo = await fetch(`${fioBaseUrl}/chain/get_info`, {
+    method: "POST",
+  }).then((res) => res.json());
+
+  const blockData = await fetch(`${fioBaseUrl}/chain/get_block`, {
+    method: "POST",
+    body: JSON.stringify({
+      block_num_or_id: chainInfo.last_irreversible_block_num,
+    }),
+  }).then((res) => res.json());
+
+  const chainData: ChainData = {
+    chain_id: chainInfo.chain_id,
+    expiration: new Date(new Date().getTime() + 120 * 1000)
+      .toISOString()
+      .split(".")[0],
+    ref_block_num: blockData.block_num & 0xffff,
+    ref_block_prefix: blockData.ref_block_prefix,
+  };
+
+  return chainData;
+};
+
+/**
+ * Broadcasts a transaction.
+ * @param rawTx The raw transaction.
+ * @param chainId The chain ID.
+ * @param account The account.
+ * @param signature The transaction signature.
+ * @returns The response of pushing the transaction.
+ */
+export const broadcastTx = async ({
+  rawTx,
+  chainId,
+  account,
+  signature,
+  fioBaseUrl,
+}: BroadcastTransactionParams): Promise<PushTransactionResponse> => {
+  const { fioTokenAbi } = await getAbiProvider(account);
+  Transactions.abiMap.set(fioTokenAbi.account_name, fioTokenAbi);
+
+  const transaction = new Transactions();
+  const { serializedContextFreeData, serializedTransaction } =
+    await transaction.serialize({
+      chainId,
+      transaction: rawTx,
+    });
+
+  const pushTransactionArgs = {
+    signatures: [signature],
+    packed_trx: Buffer.from(serializedTransaction).toString("hex"),
+    packed_context_free_data: serializedContextFreeData
+      ? Buffer.from(serializedContextFreeData).toString("hex")
+      : "",
+    compression: 0,
+  };
+
+  const pushTransactionResponse = await fetch(
+    `${fioBaseUrl}/chain/push_transaction`,
+    {
+      method: "POST",
+      body: JSON.stringify(pushTransactionArgs),
+    }
+  );
+
+  const pushTransactionJson: PushTransactionResponse =
+    await pushTransactionResponse.json();
+
+  return pushTransactionJson;
+};
+
+/**
+ * Converts an EOS public key to a FIO public key.
+ * @param eosPubKey The EOS public key.
+ * @returns The FIO public key.
+ */
+export const EOSPubKeyToFIOPubKey = (eosPubKey: string): string => {
```

Comment:
You should also copy the comments when copying the code. It makes understanding the code easier.

Reply:
Done